### PR TITLE
[feature] Implement explicit domain allows + allowlist federation mode

### PR DESCRIPTION
**Pull request by tsmethurst**
# Description

> If this is a code change, please include a summary of what you've coded, and link to the issue(s) it closes/implements.
>
> If this is a documentation change, please briefly describe what you've changed and why.

This pull request implements allowlist federation mode under a new `instance-federation-mode` config setting. To facilitate this new mode, it adds database models and functions for explicit domain allows, to complement the existing domain blocks that we already had.

closes https://github.com/superseriousbusiness/gotosocial/issues/963

## Checklist

Please put an x inside each checkbox to indicate that you've read and followed it: `[ ]` -> `[x]`

If this is a documentation change, only the first checkbox must be filled (you can delete the others if you want).

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [x] I/we have made any necessary changes to documentation.
- [x] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -3,9 +3,11 @@
 	"go.lintFlags": [
 	  "--fast"
 	],
-	"go.vetFlags": [
-		"-composites=false ."
-	],
+	"gopls": {
```
Comment (tsmethurst) [inline .vscode/settings.json:6]: This is not related to the PR, I just added it because it was annoying me. Go vet is deprecated, so setting this via gopls is the way to get rid of those annoying linter errors now :)

Comment (tsmethurst) [reply .vscode/settings.json:6]: Ahhh okay I misunderstood; yeah that's probably what's happening then!

File: internal/typeutils/internaltofrontend.go [hunk 2]
```diff
@@ -1041,32 +1041,39 @@ func (c *converter) NotificationToAPINotification(ctx context.Context, n *gtsmod
 	}, nil
 }
 
-func (c *converter) DomainBlockToAPIDomainBlock(ctx context.Context, b *gtsmodel.DomainBlock, export bool) (*apimodel.DomainBlock, error) {
+func (c *converter) DomainPermToAPIDomainPerm(
```
Comment (NyaaaWhatsUpDoc) [inline internal/typeutils/internaltofrontend.go:1044]: do we need to attach this to the type converter itself? i've been thinking recently whether in the future it might make more sense to just have `typeutils.___()` functions, the type converter itself feels like a bit of a leftover from the earlier codebase with many interfaced types. what do you think?

File: internal/gtsmodel/adminaction.go [hunk 3]
```diff
@@ -42,7 +42,7 @@ func (c AdminActionCategory) String() string {
 	case AdminActionCategoryDomain:
 		return "domain"
 	default:
-		return "unknown"
```
Comment (NyaaaWhatsUpDoc) [inline internal/gtsmodel/adminaction.go:45]: yeah fuck you `goconst` !!

File: internal/api/client/admin/domainpermissionsget.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+		return
+	}
+
+	export, errWithCode := apiutil.ParseDomainPermissionExport(c.Query(apiutil.DomainPermissionExportKey), false)
+	if errWithCode != nil {
+		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
+		return
+	}
+
+	domainPerm, errWithCode := m.processor.Admin().DomainPermissionGet(
```
Comment (NyaaaWhatsUpDoc) [inline internal/api/client/admin/domainpermissionsget.go:64]: is this meant to be the multiple form of this function you're calling? i.e. DomainPermissions instead of DomainPermission ? putting both this and the other shared code function in the same file might make this a little easier to see, i was sat looking up and down and what i thought were two instances of the same exact file

Comment (NyaaaWhatsUpDoc) [reply internal/api/client/admin/domainpermissionsget.go:64]: same goes for the deleteDomainPermission() function too, it feels a little odd with so many files scattered with single functions in them.

File: internal/api/client/admin/domainpermissioncreate.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	if err != nil {
+		apiutil.ErrorHandler(c, gtserror.NewErrorBadRequest(err, err.Error()), m.processor.InstanceGetV1)
+		return
+	}
+
+	if !importing {
+		// Single domain permission creation.
+		domainBlock, _, errWithCode := single(
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/api/client/admin/domainpermissioncreate.go:109]: wait scratch this just seen it's due to being used for creation of both allow and block permissions

File: internal/db/bundb/domain.go [hunk 6]
```diff
@@ -23,15 +23,113 @@ import (
...
 	"github.com/superseriousbusiness/gotosocial/internal/util"
 	"github.com/uptrace/bun"
 )
 
 type domainDB struct {
-	db    *DB
-	state *state.State
+	db                     *DB
+	state                  *state.State
+	instanceFederationMode string
```
Comment (NyaaaWhatsUpDoc) [inline internal/db/bundb/domain.go:36]: is there a reason this is embedded in the `domainDB{}` ? to fit with the rest of the code it feels it would make more sense to do `config.GetInstanceFederationMode()`

File: internal/gtsmodel/adminaction.go (hunk 3 above)
Comment (tsmethurst) [reply internal/gtsmodel/adminaction.go:45]: yeah!!!!

File: internal/db/bundb/domain.go [hunk 7]
```diff
@@ -157,6 +273,35 @@ func (d *domainDB) IsDomainBlocked(ctx context.Context, domain string) (bool, er
 
 		return domains, nil
 	})
+	if err != nil {
+		return false, err
+	}
+
+	// Calculate if blocked
+	// based on federation mode.
+	switch d.instanceFederationMode {
```
Comment (NyaaaWhatsUpDoc) [inline internal/db/bundb/domain.go:282]: you could just do here `mode := config.GetInstanceFederationMode(); mode` then you can still reference the variable in the error below i.e. `gtserror.Newf(..., mode)`

File: internal/typeutils/internaltofrontend.go (hunk 2 above)
Comment (tsmethurst) [reply internal/typeutils/internaltofrontend.go:1044]: Not a bad idea to move the things we can out of there, yeah... The only thing is that in many cases the typeconverter needs a db connection to do things. So I'm not sure really. I'd say this is probably something for a separate PR though, tidying up the typeconverter.

File: internal/api/client/admin/domainpermissionsget.go (hunk 4 above)
Comment (tsmethurst) [reply internal/api/client/admin/domainpermissionsget.go]: Right! Thank you, I didn't really notice it was so weird. Consolidated them into a single file now and added some comments https://github.com/superseriousbusiness/gotosocial/pull/2200/commits/c06fc41e345480336ecf9e18620cce3d5fd7c1a2

File: internal/db/bundb/domain.go (hunk 6 above)
Comment (tsmethurst) [reply internal/db/bundb/domain.go:36]: Yes, basically the goal was to avoid having to call (and lock/unlock) config every time domain blocks were checked, since that's quite a hot path. So instead it's just called once at the start and passed to the db. Happy to put it elsewhere if you prefer though!

File: internal/db/bundb/domain.go (hunk 7 above)
Comment (tsmethurst) [reply internal/db/bundb/domain.go]: Ah nice thank you, I always forget this is possible with switches! https://github.com/superseriousbusiness/gotosocial/pull/2200/commits/284b73c592090af5d3d340501bae7d6364a86910

File: internal/processing/admin/domainallow.go [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+	}
+
+	return apiDomainAllow, actionID, nil
+}
+
+func (p *Processor) domainUnallowSideEffects(
+	ctx context.Context,
+	allow *gtsmodel.DomainAllow,
+) gtserror.MultiError {
+	if config.GetInstanceFederationMode() == config.InstanceFederationModeAllowlist {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/admin/domainallow.go:222]: same goes for here, is it worth doing this check?

File: internal/processing/admin/domainallow.go [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+	}
+
+	// Prepare the domain allow to return, *before* the deletion goes through.
+	apiDomainAllow, errWithCode := p.apiDomainPerm(ctx, domainAllow, false)
+	if errWithCode != nil {
+		return nil, "", errWithCode
+	}
+
+	// Copy value of the domain allow.
+	domainAllowC := new(gtsmodel.DomainAllow)
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/admin/domainallow.go:177]: what's the reason for the copied domain allow?

File: internal/processing/admin/domainpermission.go [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+
+	// Copy the file contents into a buffer.
+	buf := new(bytes.Buffer)
+	size, err := io.Copy(buf, file)
+	if err != nil {
+		err = gtserror.Newf("error reading attachment: %w", err)
+		return nil, gtserror.NewErrorBadRequest(err, err.Error())
+	}
+
+	// Ensure we actually read something.
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/processing/admin/domainpermission.go:184]: actually with the above streaming decode change you won't need to do this anyways :p

File: internal/processing/admin/domainpermission.go [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+// true, the format will be suitable for writing out
+// to an export.
+func (p *Processor) DomainPermissionsGet(
+	ctx context.Context,
+	permissionType gtsmodel.DomainPermissionType,
+	account *gtsmodel.Account,
+	export bool,
+) ([]*apimodel.DomainPermission, gtserror.WithCode) {
+	var (
+		domainPerms *[]gtsmodel.DomainPermission
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/admin/domainpermission.go:264]: what's the reason for the ptr type here? a slice itself is a ptr to the data so just having `domainPerms []gtsmodel.DomainPermission` would serve the same purpose as below just fine i think :)

File: internal/db/bundb/domain.go (hunk 6 above)
Comment (tsmethurst) [reply internal/db/bundb/domain.go:36]: Alright got ya! Will change it :+1:

File: internal/typeutils/internaltofrontend.go (hunk 2 above)
Comment (tsmethurst) [reply internal/typeutils/internaltofrontend.go:1044]: Ah, I'd rather add it in the expected place for now and do this later in a refactoring PR, if that's alright!

File: internal/db/bundb/domain.go (hunk 6 above)
Comment (tsmethurst) [reply internal/db/bundb/domain.go]: https://github.com/superseriousbusiness/gotosocial/pull/2200/commits/5264e75a24279c30d15e1e6392a81fd88c0d267c blammo

File: internal/processing/admin/domainpermission.go [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+	// Open the provided file.
+	file, err := domainsF.Open()
+	if err != nil {
+		err = gtserror.Newf("error opening attachment: %w", err)
+		return nil, gtserror.NewErrorBadRequest(err, err.Error())
+	}
+	defer file.Close()
+
+	// Copy the file contents into a buffer.
+	buf := new(bytes.Buffer)
```
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/processing/admin/domainpermission.go:177]: Sick!

[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/processing/admin/domainpermission.go]: https://github.com/superseriousbusiness/gotosocial/pull/2200/commits/14b83fb410689addbd727cad6198280ce62a8b68

File: internal/processing/admin/domainpermission.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/processing/admin/domainpermission.go]: Nice one! Handled in streaming decode change; https://github.com/superseriousbusiness/gotosocial/pull/2200/commits/14b83fb410689addbd727cad6198280ce62a8b68

File: internal/processing/admin/domainpermission.go (hunk 11 above)
Comment (tsmethurst) [reply internal/processing/admin/domainpermission.go:248]: Ah, I figured starting with a nil pointer to a slice instead of initializing an empty slice was probably a bit cheaper. I will change it though since it's a bit weird this way.

Comment (tsmethurst) [reply internal/processing/admin/domainpermission.go]: https://github.com/superseriousbusiness/gotosocial/pull/2200/commits/0aea83593d1d4f51e7f25abfb288f06fe6c47973

File: internal/processing/admin/domainallow.go (hunk 9 above)
Comment (tsmethurst) [reply internal/processing/admin/domainallow.go:177]: Oh, good question, I think it's a relic from when we deleted these things using our generic bun delete function, which modified fields on the struct you pass into it. Will throw it out now.

Comment (tsmethurst) [reply internal/processing/admin/domainallow.go]: Yeet! https://github.com/superseriousbusiness/gotosocial/pull/2200/commits/b7d0f3b134040372a6a6de46ef2e2be61291a499

File: internal/processing/admin/domainallow.go (hunk 8 above)
Comment (tsmethurst) [reply internal/processing/admin/domainallow.go:218]: Addressed below/above (depending on how you look at it): https://github.com/superseriousbusiness/gotosocial/pull/2200/commits/b7d0f3b134040372a6a6de46ef2e2be61291a499

File: internal/processing/admin/domainallow.go [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+	}
+
+	return apiDomainAllow, actionID, nil
+}
+
+func (p *Processor) domainAllowSideEffects(
+	ctx context.Context,
+	allow *gtsmodel.DomainAllow,
+) gtserror.MultiError {
+	if config.GetInstanceFederationMode() == config.InstanceFederationModeAllowlist {
```
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/processing/admin/domainallow.go:114]: Well, in allowlist mode allows don't override blocks, instead blocks override allows, so if you:

1. Were running in blocklist mode.
2. Had an instance blocked.
3. Switched to allowlist mode.
4. Added an allow for that instance.

Then the instance would still be blocked overall. It only becomes allowed again if you add an allow AND delete the block.

File: docs/admin/domain_blocks.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+Secondly, a domain block instructs your instance to no longer make any requests to the target instance. This means:
+
+- Your instance will not deliver any messages to an instance on a blocked domain.
+- Nor will it fetch statuses, accounts, media, or emojis from that instance.
+
+## Safety concerns
+
+### Block evasion
+
+Domain blocking is not airtight. GoToSocial can guarantee that it will not serve requests from or make requests to instances on blocked domains. Unfortunately it cannot guarantee that accounts on your instance will never be visible in any way to users with accounts on blocked instances. Consider the following circumstances, all of which represent a form of [block evasion](https://en.wikipedia.org/wiki/Block_(Internet)#Evasion):
```
Comment (NyaaaWhatsUpDoc) [inline docs/admin/domain_blocks.md:29]: should this be `will not serve requests from nor make` ?

File: docs/admin/domain_blocks.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+- You've domain blocked `blocked.instance.org`. A user on `not-blocked.domain` screenshots a post of yours and sends it to someone on `blocked.instance.org`.
+- You've domain blocked `blocked.instance.org`. A user on `blocked.instance.org` visits the web view of your profile to read your public posts.
+- You've domain blocked `blocked.instance.org`. You have RSS enabled for your profile. A user from `blocked.instance.org` subscribes to your RSS feed to read your public posts.
+
+In the above cases, `blocked.instance.org` remains blocked, but users from that instance may still have other ways of seeing your posts and possibly reaching you.
+
+With this in mind, you should only ever treat domain blocking as *one layer* of your privacy onion. That is, domain blocking should be deployed alongside other layers in order to achieve a level of privacy that you are comfortable with. This ought to include things like not posting sensitive information publicly, not accidentally doxxing yourself in photos, etc.
```
Comment (NyaaaWhatsUpDoc) [inline docs/admin/domain_blocks.md:38]: there's definitely a joke to be made about leeks here :p

Comment (tsmethurst) [reply docs/admin/domain_blocks.md:38]: that might be a little farfetch'd

File: docs/admin/domain_blocks.md (hunk 14 above)
Comment (tsmethurst) [reply docs/admin/domain_blocks.md]: Changed!
